pkg/virt-controller/network/ipamclaims_test.go

```diff
@@ -0,0 +1,208 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2024 Red Hat, Inc.
+ *
+ */
+
+package network_test
+
+import (
+	"context"
+	"fmt"
+
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+
+	"github.com/golang/mock/gomock"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	networkv1 "github.com/k8snetworkplumbingwg/network-attachment-definition-client/pkg/apis/k8s.cni.cncf.io/v1"
+
+	virtv1 "kubevirt.io/api/core/v1"
+	"kubevirt.io/client-go/api"
+	"kubevirt.io/client-go/kubecli"
+
+	"kubevirt.io/kubevirt/pkg/pointer"
+	"kubevirt.io/kubevirt/pkg/virt-controller/network"
+
+	fakenetworkclient "kubevirt.io/client-go/generated/network-attachment-definition-client/clientset/versioned/fake"
+
+	ipamclaims "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1"
+	fakeipamclaimclient "github.com/k8snetworkplumbingwg/ipamclaims/pkg/crd/ipamclaims/v1alpha1/apis/clientset/versioned/fake"
+)
+
+const (
+	vmiName        = "testvmi"
+	vmiUID         = vmiName + "UID"
+	nadNetworkName = "nad_network_name"
+)
+
+var _ = Describe("CreateIPAMClaims", func() {
+	var virtClient *kubecli.MockKubevirtClient
```

Comment:
I would use a simple stub instead of gmock.

Reply:
Using it for 2 things:
1. Creating and getting NADs (as done on other places as well)
2. Creating and Listing IPAMClaims

only mocking that is done is Return().AnyTimes()
and using the fake auto generated client

Maybe one thing that can be done is changing 
```
type IPAMClaimsManager struct {
	client kubecli.KubevirtClient
}
```
to have just NetworkClient and IPAMClaimsClient instead of client
and then the unit tests wont need to mock those
do you think it is better?
or maybe you had something else in mind ?

Reply:
I don't remember what I had in mind:) Your suggestion sounds good.

Reply:
Tried it

In production code
```
type IPAMClaimsManager struct {
	networkClient    networkclient.Interface
	ipamClaimsClient ipamclaimsclient.Interface
}
```
at `NewVMIController` create ipamClaimsManager using
`ipamclaims.NewIPAMClaimsManager(clientset.NetworkClient(), clientset.IPAMClaimsClient()),`

The problem with this change, is that other unit tests that run `NewVMIController` on their flow
would now need mocks of `NetworkClient()` and `IPAMClaimsClient()` calls,
so it doesn't make sense, because we are trying to remove those calls from our test.

I checked about alternatives and didn't find yet one that is good enough.
Note please that we do need the mock only to inject the fake client, which also serves us as a cache
when testing the scenarios you asked https://github.com/kubevirt/kubevirt/pull/11410#discussion_r1579517013

Reply:
> The problem with this change, is that other unit tests that run NewVMIController on their flow
would now need mocks of NetworkClient() and IPAMClaimsClient() calls,
so it doesn't make sense, because we are trying to remove those calls from our test.

That's simply because the VMI controller constructor is being passed a mocked virtClient. 

You can avoid this by having overloading the constructor - only this new constructor will actually set the ipam claims manager. The constructor used on the existing tests will not set the ipam claims manager.

Reply:
Right thanks
What I did instead overloading the constructor is set the ipamManager outside of the constructor
in the "init" function that only production code uses, 
unless constructor overloading is preferable

Reply:
IMO overloading the constructor is clearly preferred, since it is explicit. While developing, you're forced to make a conscious choice of which constructor function to use. 

What you propose here is implicit - you're hiding setting the manager in the init function of another component. 

I won't block over this though.

Reply:
on it

Reply:
Done, will run the e2e tests on it on sunday